apis/apps/v1alpha1/opsdefinition_types.go

```diff
@@ -0,0 +1,193 @@
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package v1alpha1
+
+import (
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OpsDefinitionSpec defines the desired state of OpsDefinition
+type OpsDefinitionSpec struct {
+	// triggerPhaseChange indicates whether the operation will trigger a state change of the component.
+	// if true, will be queued for execution.
+	// +optional
+	TriggerPhaseChange bool `json:"triggerPhaseChange,omitempty"`
+
+	// componentDefinitionRefs indicates which types of componentDefinitions are supported by the operation.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +patchMergeKey=name
+	// +patchStrategy=merge,retainKeys
+	// +listType=map
+	// +listMapKey=name
+	ComponentDefinitionRefs []ComponentDefinitionRef `json:"componentDefinitionRefs" patchStrategy:"merge,retainKeys" patchMergeKey:"name"`
+
+	// parametersSchema describes the schema used for validation, pruning, and defaulting.
+	// +optional
+	ParametersSchema *ParametersSchema `json:"parametersSchema,omitempty"`
+
+	// jobSpec describes the job spec for the operation.
+	// +kubebuilder:validation:Required
+	JobSpec batchv1.JobSpec `json:"jobSpec"`
+
+	// pre-check if it meets the requirements to run the job for the operation.
+	// +optional
+	PreChecks []PreCheck `json:"preChecks,omitempty"`
+}
+
+type ComponentDefinitionRef struct {
+
+	// refer to componentDefinition name.
+	// +kubebuilder:validation:MaxLength=32
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// the account name of the component.
+	// will inject the account username and password to KB_ACCOUNT_USERNAME and KB_ACCOUNT_PASSWORD in env of the job.
+	// +optional
+	AccountName string `json:"accountName,omitempty"`
+
+	// reference the services[*].name.
+	// will map the service name and ports to KB_COMP_SVC_NAME and KB_COMP_SVC_PORT_<portName> in env of the job.
+	// portName will replace the characters '-' to '_' and convert to uppercase.
+	// +optional
+	ServiceName string `json:"serviceName,omitempty"`
+}
+
+type ParametersSchema struct {
+	// openAPIV3Schema is the OpenAPI v3 schema to use for parameter schema.
+	// supported properties types:
+	// - string
+	// - number
+	// - integer
+	// - array: only supported the item with string type.
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:Type=object
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +k8s:conversion-gen=false
+	// +optional
+	OpenAPIV3Schema *apiextensionsv1.JSONSchemaProps `json:"openAPIV3Schema,omitempty"`
+}
+
+// +kubebuilder:validation:XValidation:rule="has(self.expression) || has(self.exec)", message="at least one exists for expression and exec."
+
+type PreCheck struct {
```

Comment:
both PreCheck and Expression are two general here, pls provide them with a more specific naming